File: app-api/app/schemas/responses.py
"""Standard API response wrappers.

Maps to shared-types: ApiResponse, PaginatedResponse, ErrorResponse,
HealthCheckResponse.

Security note: Error responses NEVER include stack traces in production.
The `details` field on errors is intentionally generic.
"""

from datetime import UTC, datetime
from typing import Any, Generic, TypeVar

from pydantic import BaseModel

from app.core.pagination import calculate_total_pages
from app.schemas.base import PaginationMeta

T = TypeVar("T")


class ApiResponse(BaseModel, Generic[T]):
    """Standard API response wrapper."""

    success: bool
    data: T
    message: str | None = None
    timestamp: str
    request_id: str | None = None


class PaginatedResponse(BaseModel, Generic[T]):
    """Paginated list response."""

    success: bool = True
    data: list[T]
    pagination: PaginationMeta
    timestamp: str
    request_id: str | None = None


def make_paginated_response[T](
    data: list[T], total: int, page: int, page_size: int
) -> PaginatedResponse[T]:
    """Build PaginatedResponse from items, total, and pagination params."""
    total_pages = calculate_total_pages(total, page_size)
    return PaginatedResponse(
        success=True,
        data=data,
        pagination=PaginationMeta(
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
            has_next_page=page < total_pages,
            has_previous_page=page > 1,
        ),
        timestamp=datetime.now(UTC).isoformat(),
    )


def pagination_meta_dict(total: int, page: int, page_size: int) -> dict[str, Any]:
    """Build pagination meta as dict for embedding in custom ApiResponse payloads."""
    total_pages = calculate_total_pages(total, page_size)
    return PaginationMeta(
        total=total,
        page=page,
        page_size=page_size,
        total_pages=total_pages,
        has_next_page=page < total_pages,
        has_previous_page=page > 1,
    ).model_dump()


class ErrorDetail(BaseModel):
    """Structured error detail."""

    code: str
    message: str
    details: dict[str, Any] | None = None


class ErrorResponse(BaseModel):
    """Error response — no stack traces, no internal details."""

    success: bool = False
    error: ErrorDetail
    timestamp: str
    request_id: str | None = None


class HealthCheck(BaseModel):
    """Individual health check result."""

    name: str
    status: str  # "pass" | "warn" | "fail"
    duration: float | None = None
    message: str | None = None


class HealthCheckResponse(BaseModel):
    """Health check response."""

    status: str  # "healthy" | "degraded" | "unhealthy"
    version: str
    environment: str
    timestamp: datetime
    checks: list[HealthCheck]

File: packages/shared-types/src/api/responses.ts
// API Response types

import type { PaginationMeta } from "../utils/common";

/** Base API response */
export interface ApiResponse<T = unknown> {
  success: boolean;
  data: T;
  message?: string;
  timestamp: string;
  requestId?: string;
}

/** Paginated API response */
export interface PaginatedResponse<T> extends ApiResponse<T[]> {
  pagination: PaginationMeta;
}

/** Error response */
export interface ErrorResponse {
  success: false;
  error: ApiError;
  timestamp: string;
  requestId?: string;
}

/** API Error structure */
export interface ApiError {
  code: string;
  message: string;
  details?: Record<string, unknown>;
  validationErrors?: ValidationError[];
  stack?: string; // Only in development
}

/** Validation error detail */
export interface ValidationError {
  field: string;
  message: string;
  code: string;
  value?: unknown;
}

/** Empty success response */
export interface EmptyResponse {
  success: true;
  message?: string;
  timestamp: string;
}

/** Created resource response */
export interface CreatedResponse<T> extends ApiResponse<T> {
  /** Created resource ID */
  id: string;
  /** Resource URL */
  location?: string;
}

/** Bulk operation response */
export interface BulkOperationResponse<T = unknown> {
  success: boolean;
  /** Successfully processed items */
  succeeded: T[];
  /** Failed items with errors */
  failed: Array<{
    item: T;
    error: ApiError;
  }>;
  /** Summary */
  summary: {
    total: number;
    succeeded: number;
    failed: number;
  };
  timestamp: string;
}

/** Health check response */
export interface HealthCheckResponse {
  status: "healthy" | "degraded" | "unhealthy";
  version: string;
  environment: string;
  timestamp: string;
  checks: HealthCheck[];
}

/** Individual health check */
export interface HealthCheck {
  name: string;
  status: "pass" | "warn" | "fail";
  duration?: number; // ms
  message?: string;
}

/** Authentication response */
export interface AuthResponse {
  success: boolean;
  accessToken: string;
  refreshToken?: string;
  expiresIn: number; // seconds
  tokenType: "Bearer";
  user: {
    id: string;
    email: string;
    role: string;
    organizationId: string;
    organizationName: string;
  };
}

/** Export response */
export interface ExportResponse {
  success: boolean;
  exportId: string;
  format: "csv" | "xlsx" | "pdf" | "json";
  status: "pending" | "processing" | "completed" | "failed";
  downloadUrl?: string;
  expiresAt?: string;
}

/** Import response */
export interface ImportResponse {
  success: boolean;
  importId: string;
  status: "pending" | "validating" | "processing" | "completed" | "failed";
  progress?: {
    total: number;
    processed: number;
    succeeded: number;
    failed: number;
  };
  errors?: Array<{
    row: number;
    field: string;
    message: string;
  }>;
}

/** Webhook event */
export interface WebhookEvent<T = unknown> {
  id: string;
  type: string;
  timestamp: string;
  organizationId: string;
  data: T;
  signature?: string;
}

/** Server-Sent Event */
export interface SSEMessage<T = unknown> {
  event: string;
  id: string;
  data: T;
  retry?: number;
}

// ─────────────────────────────────────────────────────────────
// Dataset Responses
// ─────────────────────────────────────────────────────────────

import type {
  DatasetColumn,
  DatasetDataRow,
  IngestionLogEntry,
} from "../domain/dataset";
import type {
  IntegrationAuditEvent,
  IntegrationCatalogItem,
  IntegrationConnection,
  IntegrationSyncRun,
} from "../domain/integration";

/** Dataset detail response data */
export interface DatasetDetailResponse {
  id: string;
  name: string;
  status: string;
  tableName: string;
  temporalIndex: string;
  groupBy: string[];
  rowCount: number;
  lastIngestionAt: string | null;
  columns: DatasetColumn[];
}

/** Dataset data preview response */
export interface DatasetDataPreviewResponse {
  columns: string[];
  rows: DatasetDataRow[];
  /** Which columns are PII-masked */
  maskedColumns: string[];
  total: number;
}

/** Ingestion history response */
export interface IngestionHistoryResponse {
  entries: IngestionLogEntry[];
  total: number;
}

// ─────────────────────────────────────────────────────────────
// Integration Responses
// ─────────────────────────────────────────────────────────────

/** Connector catalog payload */
export interface IntegrationCatalogResponse {
  connectors: IntegrationCatalogItem[];
}

/** Connection test response payload */
export interface IntegrationConnectionTestResponse {
  ok: boolean;
  latencyMs: number;
  checkedScopes: string[];
  warnings: string[];
}

/** Integration dashboard payload */
export interface IntegrationOverviewResponse {
  organizationId: string;
  connectors: IntegrationConnection[];
  recentRuns: IntegrationSyncRun[];
  auditTrail: IntegrationAuditEvent[];
}
